### DISCUSSION TITLE
Function pointer interrupt Handler

### DISCUSSION THREAD

**Discussion by AbbottF** [rlogiacco/CircularBuffer#39] [Q&A] [answered]
Is there a way I can use the CB for handling function pointers. I’m trying to use it to queue multiple events that are pushed when either a timed or a pin interrupt occurs. 
Has anyone done this? It’s a cheap way to eliminate the need for rtos.
Thanks
Abbott
I’m usin this for an open source maple sap vacuum/pump system I’m writing. The hardware and circui boards are all operational.

**Comment by rlogiacco** [comment, marked as answer]
Yes, there’s no reason why you shouldn’t! You just declare a buffer of your type, which is going to be the function reference of your choice:

```cpp
typedef void (*func)();
CircularBuffer<func,100> functions;

void a() {}

void b() {}

void c() {}

void setup() {
  functions.push(a);
  functions.push(b);
  functions.push(c);
}
```

Please, note I haven’t compiled the code above, I just pulled it out while on the go, it’s just a reference to describe how you can do it.
Also, by varying the typedef you can establish input parameters and return values: you are not limited to functions without parameters or return values, but they must all comply with the same interface.

**Reply by AbbottHMG** [reply]
Thanks for your fast reply. I decided to add some additional features so I'm now using a class. Although the VS19 C++ editor does not show any errors, but the output shows some issues that I'm not sure how to "repair".  I'm a C# developer who hasn't coded C++ in 25 years, so please forgive my ignorance. Here's the output. 

```
Compiling debug version of 'FIFO_CB_Test' for 'ATmega2560 (Mega 2560) (Arduino Mega)'
 
cc0SSu5K.ltrans0.ltrans.o*: In function CircularBuffer<InteruptorClass, 20u, unsigned char>::push(InteruptorClass)
   (.text+0x974): undefined reference to InteruptBufferClass::interuptBuffer

Error linking for board ATmega2560 (Mega 2560) (Arduino Mega)
Debug build failed for project 'FIFO_CB_Test'
 
cc0SSu5K.ltrans0.ltrans.o*: In function InteruptBufferClass::push(InteruptorClass)
   (.text+0x9c2): undefined reference to InteruptBufferClass::interuptBuffer
   (.text+0x9c4): undefined reference to InteruptBufferClass::interuptBuffer
 
collect2.exe*: error: ld returned 1 exit status
```
[FIFO_CB.zip](https://github.com/rlogiacco/CircularBuffer/files/5721065/FIFO_CB.zip)

I'm creating a timed interrupt library for simple interrupt handling. Definitely not RTOS. NOTE-DefinedValues and Enums contain items that are related to my full project - A Maple Sap Pump/Vacuum collection system - both will be open source.
Thanks again
Abbott

**Reply by rlogiacco** [reply]
First of all, I had a quick look at your code and I don't think the problem is something in the CIrcularBuffer library, but in your code.

For what  I can understand from your code, you are trying to wrap a `CircularBuffer` inside a `InteruptBufferClass` which is going to be a specialization to handle instances of `InteruptorClass`: unless you have some specific need this is not needed. If you have a look at the `Object.ino` example, you can do something like the following:

```
#include <CircularBuffer.h>
#include "Interupt.h"

CircularBuffer<InteruptorClass*, 20> buffer;
```
When instantiated in this way, the `buffer` object is capable of handling only instances of `InteruptorClass`, as you needed.

Now, as a Java programmer myself I'm having difficulty in understanding the logic behind your code as it seems you are mixing up responsibility between the `InteruptorClass`, the `InteruptBufferClass` and their user `FIFO_CB_Test`, but I want to state something clear: the CircularBuffer does not deallocate memory, so if you instantiate a class it is your responsibility to deallocate the object memory after popping it out from the buffer. You can find additional information on the project [README](https://github.com/rlogiacco/CircularBuffer#reclaim-dynamic-memory).

Hoping this solves your doubts.

**Reply by AbbottHMG** [reply]
I did not think it was your code, but my implementation. I'll be trying to restructure my solution.
Thanks
Abbott

**Comment by AbbottHMG** [comment]
I got everything working except I seem to be screwing up memory. For the new classes, do I have to do a malloc, or something like that. I've eliminated the InteruptBuffer and am able to push 5 objects into the buffer. When I try to pop them, the first pops properly, the rest are ??? I know it's not the CB. I'm just not used to c++ and am not sure of how to proceed.
I know this is not a CB issue, but I'd appreciate a quick look.
[FIFO_CB.zip](https://github.com/rlogiacco/CircularBuffer/files/5745085/FIFO_CB.zip)

**Reply by rlogiacco** [reply]
I don't see any reference to this library in the code you have attached: I believe you'll get better and faster help on StackOverflow, here we aim at providing support to the library users.
Anyways, for what I can see at first glance your implementation is lacking in multiple aspects, like the tail wrapping around (you use the modulo operator) or the fact you are incrementing the references without checking their correctness.
Unless you are trying to learn how to write a circular buffer implementation yourself, I strongly advise to swap your implementation for this library and simplify your entire experience.

Apologies for being unable to provide additional information, but you are asking for a generic C++ code review to someone who is not a C++ expert.

**Reply by AbbottHMG** [reply]
The reference is in interupt.h:
```
// Interrupt.h

#ifndef _INTERUPT_h
#define _INTERUPT_h

#if defined(ARDUINO) && ARDUINO >= 100
	#include "arduino.h"
#else
	#include "WProgram.h"
#endif
#include "Enums.h"
#include <CircularBuffer.h>
#include "Interuptor.h"
```
Since I am one of the founders of Infragistics.com (1990), I have had to deal with others duplicating our product offerings, so I understand how you feel. I am just frustrated that my ignorance (lack of experience) with C++ has been causing me problems. C# has been my primary language.
All your examples work perfectly. I guess it's my incorporating the function pointer in the interuptor class that's screwing me up.

I will follow up with StackExchange. Once I get this working, I'll make it into a library (that requires your circular buffer). It will be a poor man's RTOS for the arduino, supporting only timed/pin interrupts but not time slicing.

Sorry if you feel that I used your time poorly. When it's up and running I'll let you know.

La ringrazio molto per il vostro tempo. Mia moglie ha vissuto a Bolongna per molti anni. Ma non parlo molto bene l'italiano.

Abbott

**Reply by AbbottHMG** [reply]
I meant StackOverflow ...

**Reply by rlogiacco** [reply]
No man, I might have been misleading: I have absolutely no problem with the creation of an alternative to this library and I'm not concerned with _duplication_, it's just I know how painful it can be to debug code on a microcontroller and also I'm truly not a C++ expert.
You can check by yourself, I'm a Java developer and I had plenty of stuff to learn to even get close to this version of the library with the most advanced additions being contributed by people much more confident with C++ then I am, like @Erlkoenig90.

If you are trying to put a series of functions in a CircularBuffer, here is the code I would use:

```cpp
#include <CircularBuffer.h>

// function signature
typedef uint8_t (*func_t)(void);
// the buffer
CircularBuffer<func_t, 10> buffer;

void setup() {
  Serial.begin(9600);
  // just an initial fill up
  buffer.push(fn1);
  buffer.push(fn2);
  buffer.push(fn3);
  buffer.push(fn4);
}

void loop() {
  if (!buffer.isEmpty()) {
    // the use of push + shift realizes a FIFO (also unshift + pop would do)
    func_t fn = buffer.shift();
    // invokes the code in the function and prints the result code
    Serial.println(fn());
  }
}

uint8_t fn1() {
  Serial.println("fn1");
  return 0;
}

uint8_t fn2() {
  Serial.println("fn2");
  return 1;
}

uint8_t fn3() {
  Serial.println("fn3");
  return 2;
}

uint8_t fn4() {
  Serial.println("fn4");
  return 3;
}
```

If, instead of functions, you want to push objects then there's an example in the library `examples` folder.

To what I can understand from the code you have submitted I'm not sure a circular buffer is what you are looking for: a CB implies the concept when too much stuff is added, some other is simply discarded... if you are going to use such a buffer to implement a RTOS, then when you try to enqueue more stuff than viable you either discard the most recently added or the oldest... Also, you are trying to add a priority concept... 

To me, it looks you are trying to create a priority queue based on a circular buffer, but such implementation would be much different because you'll have to either browse the queue by priority or populate it based on priority, meaning the backing structure cannot be an array (like the one I'm using), but rather a linked list or a double linked list (if you can trade memory for better performance).

So, it's not a matter of time, but rather a much bigger topic...

**Comment by rlogiacco** [comment]
I've had a second look at your `Interuptor.cpp` file (see my two comments below):

```cpp
InteruptorClass InteruptClass::runNextInterupt()
{
	// the following instruction retrieves the first element, also known as *head*
        // NOTE: you are not checking for buffer emptiness, meaning you might be retrieving invalid memory!
	InteruptorClass* interuptor  = InteruptClass::peek();
	if(interuptor->canRunFunction()){
		interuptor->runInterupt();
			// the following instruction extracts from the tail, also referred as *last element*
			InteruptClass::pop();
			interuptor->deleteMe();
	}
	else 
	{
	// do nothimg - where waiting for period to elapse 
	}
}
```

As you can see, you are executing one _interrupt_ but deleting another one... I would restructure the code like the following, eliminating all the additional wrapping methods:

```cpp
// define your own method to add interrupts
InteruptorClass InteruptClass::enqueue(void (*fPointer)(InteruptorClass interupt),
	EnumsClass::Interupt fifoLifoPush,
	EnumsClass::Interupt functionName,
	unsigned long period)
{
	cBuffer.unshift(new InteruptorClass(fPointer, fifoLifoPush, functionName, period)); // adds a new item to the head
}

// I would personally call this simply *next()* or *run()*
InteruptorClass InteruptClass::runNextInterupt()
{
	if(!cBuffer.isEmpty() && cBuffer.last()->canRunFunction()) { // inspects the tail
		InteruptorClass* interuptor  = cBuffer.pop(); // extracts the tail
		interuptor->runInterupt();
		// I wouldn't define a method to wrap a call to delete, you decide how objects are created and thus destroyed
		// so as I've used the *new* operator above I'm using the *delete* operator here
		delete interuptor;
	}
	else 
	{
		// if it does nothing then remove the block
	}
}
```

Your code is also hard to follow because of the names you have picked up: `Interupt` and `Interuptor` are hard to distinguish and they also define something with a different name (`InteruptClass` and `InteruptorClass`)... It was really messy to follow.
Also, I believe they are misspelt, should have double `r`... 

Finally, the code flow is totally unclear to me and, honestly, overcomplicated.

**Reply by AbbottHMG** [reply]
First of all - a Happy and a safe New Year.
I renamed interup to IHandler. SO mt IHandler manages tasks,
The purpose of Task is to hold the state of function being called. The function manages the useCount - so that the task can be left on the queue for reprocessing (a requirement of the consuming application (a Maple Pump monitoring system)).
My problem now is how to get the enqueue to point to the test function to be run.

```
void IHandlerClass::testFunction(TaskClass* task) {
	task->printMe();
}

IHandlerClass::enqueue(testFunction, EnumsClass::FifoPush, EnumsClass::CheckInTemp, 2); 
```
has compile error 
argument of type "void (*)(TaskClass *task)" is incompatible with parameter of type "void (*)(TaskClass task)"
I can;'t seem to get this right. 
I tried:
	
```
TaskClass task - new TaskClass::TaskClass(); // create dummy task

IHandlerClass::enqueue(testFunction(task), EnumsClass::FifoPush, EnumsClass::CheckInTemp, 2);
IHandlerClass::enqueue(testFunction(*task), EnumsClass::FifoPush, EnumsClass::CheckInTemp, 2);
IHandlerClass::enqueue(testFunction(&task), EnumsClass::FifoPush, EnumsClass::CheckInTemp, 2);
```
They all throw errors ???

enqueue defined as:

```
bool IHandlerClass::enqueue(void (*fPointer)(TaskClass task),
	EnumsClass::Interupt fifoLifoPush,
	EnumsClass::Interupt functionName,
	unsigned long period) {
	return cBuffer.unshift(new TaskClass(fPointer, fifoLifoPush, functionName, period )); // adds a new item to the head
}

```
Please let me know when I am annoying you too much.
FYI - This is the controller that the software is supporting:
![20200216_211653642_iOS](https://user-images.githubusercontent.com/50553899/103463777-03da3100-4cfd-11eb-86ae-c9904312a86a.png)
This is the enclosure that it will go in (to handle  8 maple saps lines with vacuum suction and pumping 300 meters away and up 100 meters to my evaporator)
![20200126_153707884_iOS](https://user-images.githubusercontent.com/50553899/103463943-20c33400-4cfe-11eb-8514-3f4c2355057a.png)
Here's the evaporator ...
![20200217_160854346_iOS](https://user-images.githubusercontent.com/50553899/103464008-b19a0f80-4cfe-11eb-80e1-a937f93420f3.png)

**Reply by rlogiacco** [reply]
> First of all - a Happy and a safe New Year.
> I renamed interup to IHandler. SO mt IHandler manages tasks,
> The purpose of Task is to hold the state of function being called. The function manages the useCount - so that the task can be left on the queue for reprocessing (a requirement of the consuming application (a Maple Pump monitoring system)).
> My problem now is how to get the enqueue to point to the test function to be run.
> 
> ```cpp
> void IHandlerClass::testFunction(TaskClass* task) {
> 	task->printMe();
> }
> 
> IHandlerClass::enqueue(testFunction, EnumsClass::FifoPush, EnumsClass::CheckInTemp, 2); 
> ```
> 
> has compile error
> argument of type "void (*)(TaskClass _task)" is incompatible with parameter of type "void (_)(TaskClass task)"
> I can;'t seem to get this right.
> I tried:
> 
> ```cpp
> TaskClass task - new TaskClass::TaskClass(); // create dummy task
> 
> IHandlerClass::enqueue(testFunction(task), EnumsClass::FifoPush, EnumsClass::CheckInTemp, 2);
> IHandlerClass::enqueue(testFunction(*task), EnumsClass::FifoPush, EnumsClass::CheckInTemp, 2);
> IHandlerClass::enqueue(testFunction(&task), EnumsClass::FifoPush, EnumsClass::CheckInTemp, 2);
> ```

The `enqueue` method accepts a _function pointer_ but you are actually trying to provide a function execution. To make it clear, the following should work:

```cpp
IHandlerClass::enqueue(testFunction, EnumsClass::FifoPush, EnumsClass::CheckInTemp, 2);
```

I provided [an example](https://github.com/rlogiacco/CircularBuffer/discussions/39#discussioncomment-256159) using a `struct` to describe the task: the call to the function is going to be performed the `runNextTask` method, as such it's in there that `testFunction(task)` is executed.

This thread is clearly no more related to the library or it's proper usage, looks like a support thread on C++ programming for which you can find better and more correct help than what you can get here.

My last generic hint: having one enumeration listing everything is useless and not different from a series of constants: the whole concept behind enumerations is to define distinct sets, so you want to have an enum for _enqueing behaviours_ and another one for _task types_.

**Reply by rlogiacco** [reply]
My previous example corrected to your new names... 

```cpp
// defines the function prototype, I'm not sure if you need a reference to the task
typedef void (*function_t)(Task task)

typedef struct {
	unsigned long executed;
	unsigned long period;
	unsigned int repetitions; // how many times this task should be repeated, once every period
	bool bubbleUp; // equivalent of you LifoPush/FifoPush: if true repeats as first
	EnumsClass type; // I wonder if this is necessary, to me this might be implied by the function
	function_t func;
} Task;


bool IHandlerClass::enqueue(function_t func, unsigned long period, unsigned int repetitions, EnumsClass type, bool bubbleUp) {
	return cBuffer.unshift(Task{millis(), period, repetitions, type, bubbleUp, func}); // adds a new item to the head
}

bool IHandlerClass::runNextTask() {
	if(!cBuffer.isEmpty() && this->canRunTask(cBuffer.last())) { // inspects the tail
 		Task task = cBuffer.pop(); // extracts the tail
		task.func(task);
		if (--task.repetitions > 0) { // decrements repetitions
			task.bubbleUp ? cBuffer.push(task) : cBuffer.unshift(task); // re-adds the item to the tail or the head depending on the bubbleUp param
		}
		return true;
	} else {
		return false;
	}
}

bool IHandlerClass::canRunTask(Task task) {
	_currentMillis = millis();
	if(millis() - task->executed >= task->period) {
		task->executed = millis();
		return true;
	} else {
		return false;
	}
}
```

**Reply by AbbottHMG** [reply]
Thanks for all your support. I was able to get everything going except for the function pointer ... which I'll follow-up on StackOverflow.

**Comment by AbbottHMG** [comment]
Thanks for your follow-up. I am dealing with a dynamic array of class object, and all the samples I see are for statically created objects. Once I can get the dynamic class array working properly, I can then use your buffer properly.
Also, once I get it working, I can simplify it and show it to you.
Thanks
Abbott

**Reply by rlogiacco** [reply]
The thing is on an MCU you have very little memory and also there is no real distinction between heap and stack as you don't have an underlying OS: in other words, you must have a very specific reason to use dynamically allocated objects as you are entering a very dangerous realm.
When you start allocating more memory than you actually have available your device will simply crash or start behaving in an unpredictable manner: there is no _out of memory_ exceptions, no warnings, nothing telling you something weird is about to happen.... it just happens. And debugging such situations is hard, to say the least.

So, I'm not telling you what to do, I'm just raising a yellow flag so that you know what you are going to have to deal with.

**Reply by AbbottHMG** [reply]
You're quite right. I am creating a static array of sufficient classes (I now call them tasks - thank you) to support the interrupt activity, As an interrupt occurs, I pick the first unused task. This way I will not commit the error of constantly allocating and de allocating memory. , InteruptClass uses the unused task ,sets its properties and pushes it onto your CB, including the function pointer. Once the task is complete, the task is reset for reuse.

**Reply by rlogiacco** [reply]
So your `Task` looks more like a `struct` than a `class`: the _behaviour_ seems given by the function, which is external, while the struct holds the _operational parameters_. The `canRunFunction` seems more related to the task processing than the task definition: if my interpretation is right you can move that code into your _task runner_ (the `InterruptClass`, for what I understood).

I took the liberty to add a `repetitions` property to the task which re-insert the task in its position, for the sake of making the example clearer in terms of responsibility.
```cpp
// defines the function prototype, I'm not sure if you need a reference to the task
typedef void (*function_t)(Task task)

typedef struct {
	unsigned long executed;
	unsigned long period;
	unsigned int repetitions; // how many times this task should be repeated, once every period
	function_t func;
} Task;


bool InteruptClass::enqueue(const function_t func, const unsigned long period, const unsigned int repetitions) {
	return cBuffer.unshift(Task{millis(), period, repetitions, func}); // adds a new item to the head
}

bool InteruptClass::runNextTask() {
	if(!cBuffer.isEmpty() && this->canRunTask(cBuffer.last())) { // inspects the tail
 		Task task = cBuffer.pop(); // extracts the tail
		task.func(task);
		if (--task.repetitions > 0) { // decrements repetitions
			cBuffer.push(task); // re-adds the item to the tail, but you might want to add it to the head, depending on the behaviour you want to achieve
		}
		return true;
	} else {
		return false;
	}
}

bool InteruptClass::canRunTask(Task task) {
	_currentMillis = millis();
	if(millis() - task->executed >= task->period) {
		task->executed = millis();
		return true;
	} else {
		return false;
	}
}
```

The great advantage of using a struct is the whole memory necessary to hold the circular buffer is allocated when the circular buffer is initialized and can be checked at compile time.

**Comment by proddy** [comment]
@AbbottHMG I did something similar, having a dynamically growing array in ESP8266's heap of class/structs. I used a modified version of `array` https://github.com/muwerk/ustd since std::array is static. Combined with CircularBuffer it's a good match. I also tried out std::queue and std::dequeue but CircularBuffer beats them every time.

**Reply by AbbottHMG** [reply]
Thanks. I'm looking at it now.

**Comment by rlogiacco** [comment]
It would be nice if you mark the answer which best responds to your question.

Thank you

**Comment by AbbottHMG** [comment]
This is the working test version of my task manager. The task manger controls the CB and the task array that it points to. It also fills the CB with objects in the array. The ino loop empties the buffer.
[CBTest.zip](https://github.com/rlogiacco/CircularBuffer/files/5791456/CBTest.zip)
Let me know if there is anything else you might need from me.
I coulod not get it to work without using the externs. Probably some C++ binding issue.
Thanks.

**Reply by rlogiacco** [reply]
I'm only asking you to select here the answer which best responds to your question.
